lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -333,14 +338,17 @@ def get_paginated_obj(
         Database.dump_query(query, current_app.logger)
 
         items = query.all()
+        raw = raw_params.query.copy()
         next_offset = offset + len(items)
         if next_offset < total_count:
-            json["offset"] = next_offset
+            json["offset"] = str(next_offset)
+            raw["offset"] = str(next_offset)
```

Comment:
I understand, but I also didn't want the parameters to be inconsistent with the `next_url`, because I think *that's* confusing. The client can adjust the offset anywhere they want; and really my expectation is that with random access the client doesn't even actually *care* because it'll compute an offset based on the `limit` and `total` values... but omitting it didn't make much sense.

You are however correct about the "last page" glitch, and I really don't know how I'd *want* to deal with that. Reporting the "current" `offset` would avoid that, but I still don't like that we're disconnecting the `next_url` and `parameters`. Right now I'm not sure which is worse or whether either is better and/or maybe less confusing. 😦

Reply:
> I also didn't want the parameters to be inconsistent with the `next_url`

Understood.  I would like to present the parameters as independent of `next_url`...in the hope that that would alleviate any confusion.

> The client can adjust the offset anywhere they want [...] based on the `limit` and `total` values

This is, of course, true.  I wonder if the Dashboard should be doing that instead of deconstructing the `next_url` value....

In that sense, returning the deconstructed URL is kind of silly, because it contains stuff that the client already specified in the request, combined with stuff that the client can calculate...so, we're getting into "convenience territory" here, and that's hard to evaluate without knowing how it will be used.  We currently have one client and it apparently wants the _next_ offset, so I s'pose we could return that.

As for the last page, I would just return an `"offset"` value of `"total"`.  After all, that _is_ the offset to make the next request from if something new is uploaded....

Reply:
> This is, of course, true. I wonder if the Dashboard should be doing that instead of deconstructing the `next_url` value....

Well, that's also true. All the filter and sort parameters were initially specified, presumably from an object given to the query API, and all we really need to do is keep that and, on each scroll, calculate the proper `offset`, which we're doing anyway. That was kinda what I thought/expected initially, but I got carried away by your suggestion. This *does* mean that the client doesn't *have* to keep that original query object, because we do that for them. If they're scrolling sequentially, `next_url` is all they need and, if not, they have to override the `offset` anyway.

All of which kinda makes me wonder why I bothered, while I'd also feel bad about throwing this away. Hmm.

> In that sense, returning the deconstructed URL is kind of silly, because it contains stuff that the client already specified in the request, combined with stuff that the client can calculate...so, we're getting into "convenience territory" here, and that's hard to evaluate without knowing how it will be used. We currently have one client and it apparently wants the _next_ offset, so I s'pose we could return that.
> 
> As for the last page, I would just return an `"offset"` value of `"total"`. After all, that _is_ the offset to make the next request from if something new is uploaded....

Reply:
> kinda makes me wonder why I bothered

I think the answer is tied up in the Dashboard implementation (for which I largely blame Javascript and Redux):  the stuff used to create the request came from some other function somewhere else, and the context isn't necessarily available here where we're processing the response (and setting up the next request).  So, we're working with what we have, and would it be nice if the response made that easy....